slatedb/src/retention_iterator.rs

```diff
@@ -0,0 +1,802 @@
+            filter_tombstone,
+            start_timestamp,
+            buffer: RetentionBuffer::new(),
+            total_bytes_processed: 0,
+        })
+    }
+
+    /// Applies retention filtering to a collection of versions for the same key
+    ///
+    /// This function implements the core retention logic:
+    /// - Always preserves the latest version (highest sequence number)
+    /// - Filters out older versions that exceed the retention period
+    /// - Uses `create_ts` to determine if a version should be retained
+    fn apply_retention_filter(
+        versions: BTreeMap<Reverse<u64>, RowEntry>,
+        current_timestamp: i64,
+        retention_time: Duration,
+        filter_tombstone: bool,
+    ) -> BTreeMap<Reverse<u64>, RowEntry> {
+        let mut filtered_versions = BTreeMap::new();
+        for (idx, (_, entry)) in versions.into_iter().enumerate() {
+            // always keep the latest version (idx == 0), for older versions, check if they are within retention window
+            let in_retention_window = entry
+                .create_ts
+                .map(|create_ts| create_ts + (retention_time.as_secs() as i64) >= current_timestamp)
+                .unwrap_or(true);
+            let should_keep = idx == 0 || in_retention_window;
+            if !should_keep {
+                continue;
+            }
+
+            // filter out any expired entries -- eventually we can consider
+            // abstracting this away into generic, pluggable compaction filters
+            // but for now we do it inline
+            let entry = match entry.expire_ts.as_ref() {
+                Some(expire_ts) if *expire_ts <= current_timestamp => {
+                    // insert a tombstone instead of just filtering out the
+                    // value in the iterator because this may otherwise "revive"
+                    // an older version of the KV pair that has a larger TTL in
+                    // a lower level of the LSM tree
+                    RowEntry {
+                        key: entry.key,
+                        value: Tombstone,
+                        seq: entry.seq,
+                        expire_ts: None,
+                        create_ts: entry.create_ts,
+                    }
+                }
+                _ => entry,
+            };
+
+            filtered_versions.insert(Reverse(entry.seq), entry);
+        }
+
+        // remove the tombstones in the tail
+        if filter_tombstone {
+            while filtered_versions
+                .iter()
+                .last()
+                .map(|(_, entry)| entry.value.is_tombstone())
+                .unwrap_or(false)
+            {
+                filtered_versions.pop_last();
+            }
+        }
+
+        filtered_versions
+    }
+
+    pub(crate) fn total_bytes_processed(&self) -> u64 {
+        self.total_bytes_processed
+    }
+}
+
+#[async_trait]
+impl<T: KeyValueIterator> KeyValueIterator for RetentionIterator<T> {
```

Comment:
Curious about the choice to make this a `KeyValueIterator` vs. a `FilterPredicate` for `FilterIterator`?